### DISCUSSION TITLE
Nested Components Loop of Model N+1 - Need More Best Practices Examples

### DISCUSSION THREAD

**Discussion by sergebenard** [livewire/livewire#5503] [1. Help] [answered]
Hello!

I'm working on a "like/dislike" system, and I'm trying to improve the performance of my app.

This is how my app is built:

- **Reports.Index Component** - Passes **ONLY** the Report Model's ID to:
  - **Reports.Card Component** - Loads up the Report as well as a relationship of the Report.
  - This more than doubles the amount of Queries when compared to passing the Report's Model to the Card Component, on top of creating an N+1 problem.

Everything works great if I pass the Report Model to the Reports.Card Component - but the payload for each Card is 5Kb+, and the Lighthouse tests show me at a very concerning 68%. So, I thought there must be a better way.

I'd like to keep the Reports.Card Component, because I want to isolate the concerns of the Card's presentation and functionality. But is that even possible without the N+1? Are the solutions simple?

I've read from [Michael Rubel's Github](https://github.com/michael-rubel/livewire-best-practices) that in order to have Performant Livewire, [I should not pass objects to the `mount()` method](https://github.com/michael-rubel/livewire-best-practices#-the-golden-rule-of-performant-livewire).

Additionally, Michael's Best Practices Github [there are examples of how to use Route Model Binding](https://github.com/michael-rubel/livewire-best-practices/blob/main/Examples/route-model-binding.md), and so I decided to load each individual Report Card in this way. However, once I've done this, I found **_it creates an N+1 problem_**.

This is what I've got - I've simplified my code, and I've confirmed that what you see here to creates that very same N+1 problem:

`Reports.Index's render() Function in Livewire Component Class:`
```php
    public function render() {
        $this->useCachedRows();

        return view('livewire.reports.index', [
            'reports' => Report::select(['id'])
                            ->latest()
                            ->paginate( $this->paginateAmount )
        ]);
    }
```

`Reports.Index Livewire Blade File:`
```blade
    <div class="grid grid-cols-1 sm:grid-cols-2 lg:grid-cols-3 lg:gap-5" > <!-- Grid for cards -->
        @forelse ($reports as $report)
            <livewire:reports.components.card :report="$report->id" :wire:key="'report-' . $report->id .'-' . time()" />

        @empty
            <div class="sm:col-span-2 lg:col-span-3">
                @lang('reports-index.emptyReportsMessageText')
            </div>
        @endforelse
    </div>
```

`Reports.Card Livewire Component Class:`
```php
namespace App\Http\Livewire\Reports\Components;

use App\Models\Report;
use App\Models\Ruling;
use Livewire\Component;
use Illuminate\Database\Eloquent\Builder;

class Card extends Component
{
    public $report;

    public function mount(Report $report)
    {
        $this->fill(['report' => $report->load([
                    'infractions' => function( $query ) {
                        $query->select(['id', 'short_name', 'icon'])
                            ->orderBy('short_name');
                    }
                ]);
    }

    public function render()
    {

        return view('livewire.reports.components.card');
    }
}
```
Having done all of that, I am not even sure I'm doing this correctly.

I'd like to have a good example to look at so I know where I'm going wrong.

Can someone help? At this point, any help is very much appreciated!

Serge

**Comment by joshhanley** [comment, marked as answer]
@sergebenard I'm not sure why Micheal Rubel doesn't recommend passing objects via `mount()`. I do this all the time without any issues. I would agree if he said not to add models to public properties. But passing through `mount` is fine in my opinion.

I just answered a similar question on another discussion, have a look at it, and see if it might help you with your N+1 problem. https://github.com/livewire/livewire/discussions/5493

Hope this helps!

**Reply by sergebenard** [reply]
This is great, thank you for your help!

I've switched over to passing a Model object to the Reports.Card component, then eager loading the Report model into Report.Card's `mount()` method, and within the `mount()` method I use the `$this->fill()` method to populate Report.Card's public `$reports` variable.

Not only did it solve the N+1 problem, but it also bumped my Lighthouse performance up 12%!

Thanks lots!

**Reply by michael-rubel** [reply]
@joshhanley @sergebenard really! There's no technical necessity to avoid objects in mount.

Though I added that from the start because people often do assignments in the old PHP 7.4-construct way:

```php
public $user;
public $post;

public function mount(User $user, Post $post)
{
    $this->user = $user;
    $this->post = $post;
}
```

And this ☝️ of course hurts Livewire performance.

I've updated the Best Practices to avoid this confusion! 😉

**Reply by joshhanley** [reply]
@michael-rubel ah yeah, makes sense 🙂
